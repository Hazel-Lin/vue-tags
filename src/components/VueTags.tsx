import type { PropType } from 'vue'
import { defineComponent, nextTick, ref } from 'vue'
import type { Tag } from '../types/vueTags'
import SingleTag from './SingleTag'

const VueTags = defineComponent({
  name: 'VueTags',
  props: {
    tags: {
      type: Array as PropType<Tag[]>,
      required: true,
    },
    handleAddition: {
      type: Function,
      default: Function as PropType<(tag: Tag) => void>,
    },
    allowAdditionFromPaste: {
      type: Boolean as PropType<boolean>,
      default: true,
    },
    placeholder: {
      type: String as PropType<string>,
      default: 'Add new tag',
    },
    handleDelete: {
      type: Function as PropType<(index: number, event: MouseEvent) => void>,
      default: () => {},
    },
    handleClearAll: {
      type: Function as PropType<() => void>,
      default: () => {},
    },
    handleChangeTag: {
      type: Function as PropType<(index: number, newText: string) => void>,
      default: () => {},
    },
    handleInputChange: {
      type: Function as PropType<(value: string, event: Event) => void>,
      default: () => {},
    },
    handleInputBlur: {
      type: Function as PropType<(value: string, event: FocusEvent) => void>,
      default: () => {},
    },
    classNames: {
      type: Object as PropType<Record<string, string>>,
      default: () => ({}),
    },
  },
  setup(props) {
    const {
      allowAdditionFromPaste,
      placeholder,
      handleClearAll,
    } = props
    const tags = ref(props.tags)
    const inputText = ref()
    const newText = ref()
    const currentEditIndex = ref(-1)
    const inputRef = ref()
    const tagInputRef = ref()

    const isExist = ref(false)

    // TODO 优化 可以自定义类名
    const clearAllClass = 'cursor-pointer p-2.5 bg-red-500 text-white rounded border-none ml-2'

    const onAdd = (value: string) => {
      const newTag: Tag = { id: String(tags.value.length + 1), name: value }

      isExist.value = tags.value.some((tag: Tag) => tag.name === value)
      !isExist.value && (tags.value.push(newTag) && (inputText.value = ''))
    }
    /**
     * @description 删除标签
     * @param index 标签索引
     * @param event 事件对象
     */
    function handleDeleteTag(index: number, event: MouseEvent) {
      event.preventDefault()
      event.stopPropagation()

      props?.handleDelete?.(index, event)
      tags.value.splice(index, 1)
    }
    function handleClick() {
      handleClearAll?.()
      tags.value = []
      inputText.value = ''
    }
    const handleChange = (event: Event) => {
      const value = (event.target as HTMLInputElement).value

      if (props.handleInputChange) {
        props.handleInputChange(value, event)
      }
    }
    const handleKeyDown = (event: KeyboardEvent) => {
      const value = (event.target as HTMLInputElement).value
      if (event.key === 'Enter') {
        if (value) {
          currentEditIndex.value !== -1 ? props?.handleChangeTag?.(currentEditIndex.value, value) : onAdd(value)
        }
        currentEditIndex.value = -1
      }
    }
    const handleBlur = (event: FocusEvent) => {
      const value = (event.target as HTMLInputElement).value
      if (props.handleInputBlur) {
        props.handleInputBlur(value, event)
      }
      currentEditIndex.value = -1
    }

    const handlePaste = (event: ClipboardEvent) => {
      if (!allowAdditionFromPaste) {
        return
      }
      // 解决粘贴后，输入框的值不清空的问题
      event.preventDefault()

      const clipboardData = event.clipboardData || (window as any).clipboardData
      const clipboardText = clipboardData.getData('text')
      if (clipboardText) {
        inputText.value = ''
        currentEditIndex.value !== -1 ? props?.handleChangeTag?.(currentEditIndex.value, clipboardText) : onAdd(clipboardText)
      }
      currentEditIndex.value = -1
    }
    function handleClickTag(index: number, text: string) {
      currentEditIndex.value = index
      // 自动 focus 函数
      nextTick(() => {
        tagInputRef.value?.focus()
        newText.value = text
      })
    }
    return () => (
      <div>
        <div class="flex">
          {tags.value.map((tag: Tag, index: number) => (
            <div>
              {currentEditIndex.value === index
                ? (
                  <input
                    ref={tagInputRef}
                    class="h-8"
                    type="text"
                    value={newText.value}
                    onChange={handleChange}
                    onKeydown={handleKeyDown}
                    onBlur={handleBlur}
                    onPaste={handlePaste}
                  />
                  )
                : (
                  <SingleTag
                    key={tag.id}
                    id={tag.id}
                    name={tag.name}
                    onDelete={(event: MouseEvent) => handleDeleteTag(index, event)}
                    onTagClicked={() => handleClickTag(index, tag.name)}
                  />
                  )}
            </div>
          ))}

        </div>
        <div class="mt-2 flex">
          <input
            class="h-8"
            ref={inputRef.value}
            type="text"
            value={inputText.value}
            placeholder={placeholder}
            onChange={handleChange}
            onKeydown={handleKeyDown}
            onBlur={handleBlur}
            onPaste={handlePaste}
            data-automation="input"
          />
          { tags.value.length > 0 && (
            <button
              onClick={handleClick}
              class={clearAllClass}
            >
              一键清空
            </button>
          )}

        </div>
      </div>
    )
  },

})

export default VueTags
