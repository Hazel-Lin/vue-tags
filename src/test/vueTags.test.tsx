import { describe, expect, it, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/vue'
import { mount } from '@vue/test-utils'
import type { Tag } from '../types/vueTags'
import VueTags from '../components/VueTags'

describe('test VueTags', () => {
  const sampleTags: Tag[] = [
    { id: '1', name: 'Thailand' },
    { id: '2', name: 'India' },
  ]

  it.skip('should render correctly', () => {
    const { getByText } = render(VueTags, {
      props: {
        tags: sampleTags,
      },
    })
    // sampleTags 获取到的文本是否存在
    for (const tag of sampleTags) {
      expect(getByText(tag.name)).toBeTruthy()
    }
    // 输入框是否存在
    expect(screen.getByRole('textbox')).toBeTruthy()
    expect(screen.getByPlaceholderText('Add new tag')).toBeTruthy()
    // 一键清空按钮是否存在
    expect(getByText('一键清空')).toBeTruthy()
  })

  describe('when readOnly is true', () => {
    it.skip('should not render input', () => {
      const container = render(VueTags, {
        props: {
          tags: sampleTags,
          readOnly: true,
        },
      })
      // 当 readOnly 为 true 时，输入框不存在
      expect(container.queryByTestId('input')).toBeNull()
    })
    it('should not edit tag', async () => {
      const container = render(VueTags, {
        props: {
          tags: sampleTags,
          readOnly: true,
          editable: true,
        },
      })
      // 点击某一个 tag 后，编辑框不存在
      await fireEvent.click(container.getByText('Thailand'))
      expect(container.queryByTestId('tag-edit')).toBeNull()
    })
    it('should not delete tag', () => {
      const container = render(VueTags, {
        props: {
          tags: sampleTags,
          readOnly: true,
        },
      })
      // 删除按钮不存在
      expect(container.queryByTestId('tag-delete')).toBeNull()
    })
    it('should not drag tag', async () => {
      const { getByText, container } = render(VueTags, {
        props: {
          tags: sampleTags,
          readOnly: true,
          allowDrag: true,
        },
      })
      // 不存在这个样式
      expect(container.querySelectorAll('.cursor-move').length).toEqual(0)
      const dest = getByText('Thailand')

      // TODO 模拟拖拽
      await fireEvent.dragStart(dest)
      await fireEvent.dragEnter(dest)
      await fireEvent.dragOver(dest)
      await fireEvent.drop(dest)
      await fireEvent.dragEnd(dest)
    })
  })

  // it('should adds a new tag', async () => {
  //   const wrapper = mount(VueTags, {
  //     props: {
  //       tags: sampleTags,
  //       handleAddition: vi.fn(),
  //     },
  //   })

  //   const input = wrapper.find('input[data-automation="input"]')
  //   await input.setValue('NewTag')
  //   await input.trigger('keydown.enter')

  //   expect(wrapper.props().handleAddition).toHaveBeenCalled()
  // })

  // it('deletes a tag', async () => {
  //   const handleDelete = vi.fn()
  //   const wrapper = mount(VueTags, {
  //     props: {
  //       tags: sampleTags,
  //       handleDelete,
  //     },
  //   })

  //   const deleteButtons = wrapper.findAll('.single-tag .delete-button')
  //   await deleteButtons[0].trigger('click')

  //   expect(handleDelete).toHaveBeenCalledWith(0, expect.any(MouseEvent))
  // })

  // it('clears all tags', async () => {
  //   const handleClearAll = vi.fn()
  //   const wrapper = mount(VueTags, {
  //     props: {
  //       tags: sampleTags,
  //       handleClearAll,
  //       clearAll: true,
  //     },
  //   })

  //   const clearButton = wrapper.find('button.clear-button')
  //   await clearButton.trigger('click')

  //   expect(handleClearAll).toHaveBeenCalled()
  //   expect(wrapper.props().tags).toHaveLength(0)
  // })

  // it('edits a tag', async () => {
  //   const handleChangeTag = vi.fn()
  //   const wrapper = mount(VueTags, {
  //     props: {
  //       tags: sampleTags,
  //       editable: true,
  //       handleChangeTag,
  //     },
  //   })

  //   const editInputs = wrapper.findAll('input.h-8')
  //   await editInputs[0].setValue('EditedTag')
  //   await editInputs[0].trigger('blur')

  //   expect(handleChangeTag).toHaveBeenCalledWith(0, 'EditedTag')
  // })
})
