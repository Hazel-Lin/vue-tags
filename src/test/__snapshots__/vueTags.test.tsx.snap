// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test vueTags > should render correctly 1`] = `
{
  "baseElement": <body>
    <div>
      <div
        class="grid gap-sm"
      >
        <!---->
        <div
          class="flex gap-2"
        >
          <div
            class="flex"
          >
            
            <div>
              <div
                class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
              >
                <span
                  class="mr-2"
                >
                  Thailand
                </span>
                <span
                  class="cursor-pointer"
                >
                  x
                </span>
              </div>
            </div>
            <div>
              <div
                class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
              >
                <span
                  class="mr-2"
                >
                  India
                </span>
                <span
                  class="cursor-pointer"
                >
                  x
                </span>
              </div>
            </div>
            <div>
              <div
                class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
              >
                <span
                  class="mr-2"
                >
                  Vietnam
                </span>
                <span
                  class="cursor-pointer"
                >
                  x
                </span>
              </div>
            </div>
            <div>
              <div
                class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
              >
                <span
                  class="mr-2"
                >
                  Turkey
                </span>
                <span
                  class="cursor-pointer"
                >
                  x
                </span>
              </div>
            </div>
            
          </div>
          <!---->
        </div>
        <!---->
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="grid gap-sm"
    >
      <!---->
      <div
        class="flex gap-2"
      >
        <div
          class="flex"
        >
          
          <div>
            <div
              class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
            >
              <span
                class="mr-2"
              >
                Thailand
              </span>
              <span
                class="cursor-pointer"
              >
                x
              </span>
            </div>
          </div>
          <div>
            <div
              class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
            >
              <span
                class="mr-2"
              >
                India
              </span>
              <span
                class="cursor-pointer"
              >
                x
              </span>
            </div>
          </div>
          <div>
            <div
              class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
            >
              <span
                class="mr-2"
              >
                Vietnam
              </span>
              <span
                class="cursor-pointer"
              >
                x
              </span>
            </div>
          </div>
          <div>
            <div
              class="ml-2 bg-blue-400 text-white px-1.5 py-2 rounded"
            >
              <span
                class="mr-2"
              >
                Turkey
              </span>
              <span
                class="cursor-pointer"
              >
                x
              </span>
            </div>
          </div>
          
        </div>
        <!---->
      </div>
      <!---->
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
